File: admin_platform/cpp/src/handlers/kyc_handler.cpp
/**
 * TigerWallet Admin Platform - C++ KYC Handler Implementation
 * High-performance KYC verification handler
 */

#include "../include/handlers/kyc_handler.h"
#include "../include/database.h"
#include "../include/redis_client.h"
#include <iostream>
#include <sstream>

namespace tiger {
namespace handlers {

KYCHandler::KYCHandler(std::shared_ptr<Database> db, std::shared_ptr<RedisClient> redis)
    : db_(db), redis_(redis) {}

nlohmann::json KYCHandler::get_kyc_submissions(const nlohmann::json& request) {
    int page = request.value("page", 1);
    int limit = request.value("limit", 20);
    std::string status = request.value("status", "");
    int level = request.value("level", 0);
    
    std::stringstream query;
    query << "SELECT id, user_id, level, document_type, document_number, "
          << "first_name, last_name, country, address, status, reject_reason, "
          << "reviewed_by, reviewed_at, created_at FROM kyc_submissions WHERE 1=1";
    
    if (!status.empty()) {
        query << " AND status = '" << db_->escape_string(status) << "'";
    }
    
    if (level > 0) {
        query << " AND level = " << level;
    }
    
    query << " ORDER BY created_at DESC";
    query << " LIMIT " << limit << " OFFSET " << ((page - 1) * limit);
    
    auto result = db_->execute(query.str());
    
    if (!result) {
        return nlohmann::json::object({
            {"error", "Failed to fetch KYC submissions"},
            {"success", false}
        });
    }
    
    nlohmann::json submissions = nlohmann::json::array();
    
    for (int i = 0; i < result->rows(); ++i) {
        submissions.push_back(nlohmann::json::object({
            {"id", result->get_value(i, "id")},
            {"user_id", result->get_value(i, "user_id")},
            {"level", std::stoi(result->get_value(i, "level"))},
            {"document_type", result->get_value(i, "document_type")},
            {"document_number", result->get_value(i, "document_number")},
            {"first_name", result->get_value(i, "first_name")},
            {"last_name", result->get_value(i, "last_name")},
            {"country", result->get_value(i, "country")},
            {"address", result->get_value(i, "address")},
            {"status", result->get_value(i, "status")},
            {"reject_reason", result->get_value(i, "reject_reason")},
            {"reviewed_by", result->get_value(i, "reviewed_by")},
            {"reviewed_at", result->get_value(i, "reviewed_at")},
            {"created_at", result->get_value(i, "created_at")}
        }));
    }
    
    // Get total count
    std::stringstream count_query;
    count_query << "SELECT COUNT(*) as total FROM kyc_submissions WHERE 1=1";
    if (!status.empty()) {
        count_query << " AND status = '" << db_->escape_string(status) << "'";
    }
    if (level > 0) {
        count_query << " AND level = " << level;
    }
    auto count_result = db_->execute(count_query.str());
    int total = count_result && count_result->rows() > 0 ? 
        std::stoi(count_result->get_value(0, "total")) : 0;
    
    return nlohmann::json::object({
        {"success", true},
        {"data", submissions},
        {"meta", nlohmann::json::object({
            {"page", page},
            {"limit", limit},
            {"total", total},
            {"total_pages", (total + limit - 1) / limit}
        })}
    });
}

nlohmann::json KYCHandler::get_kyc(const std::string& kyc_id) {
    std::stringstream query;
    query << "SELECT id, user_id, level, document_type, document_number, "
          << "first_name, last_name, country, address, status, reject_reason, "
          << "reviewed_by, reviewed_at, created_at, documents "
          << "FROM kyc_submissions WHERE id = '" << db_->escape_string(kyc_id) << "'";
    
    auto result = db_->execute(query.str());
    
    if (!result || result->rows() == 0) {
        return nlohmann::json::object({
            {"error", "KYC submission not found"},
            {"success", false}
        });
    }
    
    return nlohmann::json::object({
        {"success", true},
        {"data", nlohmann::json::object({
            {"id", result->get_value(0, "id")},
            {"user_id", result->get_value(0, "user_id")},
            {"level", std::stoi(result->get_value(0, "level"))},
            {"document_type", result->get_value(0, "document_type")},
            {"document_number", result->get_value(0, "document_number")},
            {"first_name", result->get_value(0, "first_name")},
            {"last_name", result->get_value(0, "last_name")},
            {"country", result->get_value(0, "country")},
            {"address", result->get_value(0, "address")},
            {"status", result->get_value(0, "status")},
            {"reject_reason", result->get_value(0, "reject_reason")},
            {"reviewed_by", result->get_value(0, "reviewed_by")},
            {"reviewed_at", result->get_value(0, "reviewed_at")},
            {"created_at", result->get_value(0, "created_at")},
            {"documents", result->get_value(0, "documents")}
        })}
    });
}

nlohmann::json KYCHandler::approve_kyc(const std::string& kyc_id, const std::string& admin_id, const std::string& notes) {
    db_->begin();
    
    try {
        // Update KYC status
        std::stringstream query;
        query << "UPDATE kyc_submissions SET status = 'approved', "
              << "reviewed_by = '" << db_->escape_string(admin_id) << "', "
              << "reviewed_at = NOW(), "
              << "notes = '" << db_->escape_string(notes) << "' "
              << "WHERE id = '" << db_->escape_string(kyc_id) << "'";
        
        auto result = db_->execute(query.str());
        
        if (!result) {
            db_->rollback();
            return nlohmann::json::object({
                {"error", "Failed to approve KYC"},
                {"success", false}
            });
        }
        
        // Get user_id from KYC submission
        std::stringstream user_query;
        user_query << "SELECT user_id FROM kyc_submissions WHERE id = '" << db_->escape_string(kyc_id) << "'";
        auto user_result = db_->execute(user_query.str());
        
        if (user_result && user_result->rows() > 0) {
            std::string user_id = user_result->get_value(0, "user_id");
            
            // Update user KYC status
            std::stringstream update_user;
            update_user << "UPDATE users SET kyc_status = 'approved', "
                        << "kyc_level = (SELECT level FROM kyc_submissions WHERE id = '" 
                        << db_->escape_string(kyc_id) << "') "
                        << "WHERE id = '" << db_->escape_string(user_id) << "'";
            db_->execute(update_user.str());
            
            // Invalidate cache
            if (redis_) {
                redis_->del("kyc:" + kyc_id);
                redis_->del("user:" + user_id);
            }
        }
        
        db_->commit();
        
        return nlohmann::json::object({
            {"success", true},
            {"message", "KYC approved successfully"}
        });
        
    } catch (const std::exception& e) {
        db_->rollback();
        return nlohmann::json::object({
            {"error", e.what()},
            {"success", false}
        });
    }
}

nlohmann::json KYCHandler::reject_kyc(const std::string& kyc_id, const std::string& admin_id, const std::string& reason) {
    db_->begin();
    
    try {
        // Update KYC status
        std::stringstream query;
        query << "UPDATE kyc_submissions SET status = 'rejected', "
              << "reviewed_by = '" << db_->escape_string(admin_id) << "', "
              << "reviewed_at = NOW(), "
              << "reject_reason = '" << db_->escape_string(reason) << "' "
              << "WHERE id = '" << db_->escape_string(kyc_id) << "'";
        
        auto result = db_->execute(query.str());
        
        if (!result) {
            db_->rollback();
            return nlohmann::json::object({
                {"error", "Failed to reject KYC"},
                {"success", false}
            });
        }
        
        // Get user_id from KYC submission
        std::stringstream user_query;
        user_query << "SELECT user_id FROM kyc_submissions WHERE id = '" << db_->escape_string(kyc_id) << "'";
        auto user_result = db_->execute(user_query.str());
        
        if (user_result && user_result->rows() > 0) {
            std::string user_id = user_result->get_value(0, "user_id");
            
            // Update user KYC status
            std::stringstream update_user;
            update_user << "UPDATE users SET kyc_status = 'rejected' "
                        << "WHERE id = '" << db_->escape_string(user_id) << "'";
            db_->execute(update_user.str());
            
            // Invalidate cache
            if (redis_) {
                redis_->del("kyc:" + kyc_id);
                redis_->del("user:" + user_id);
            }
        }
        
        db_->commit();
        
        return nlohmann::json::object({
            {"success", true},
            {"message", "KYC rejected successfully"}
        });
        
    } catch (const std::exception& e) {
        db_->rollback();
        return nlohmann::json::object({
            {"error", e.what()},
            {"success", false}
        });
    }
}

} // namespace handlers
} // namespace tiger

File: admin_backend/internal/handlers/kyc_handler.go
package handlers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"admin_backend/internal/models"
	"admin_backend/pkg/database"

	"github.com/gin-gonic/gin"
	"github.com/redis/go-redis/v9"
)

// NotificationService handles sending notifications to users
type NotificationService struct {
	redis *redis.Client
}

// NewNotificationService creates a new notification service
func NewNotificationService(redisClient *redis.Client) *NotificationService {
	return &NotificationService{redis: redisClient}
}

// SendKYCNotification sends a KYC status notification to a user via Redis queue
func (s *NotificationService) SendKYCNotification(userID uint, status, message string) error {
	// Store notification in Redis queue for processing by notification service
	key := fmt.Sprintf("notifications:user:%d", userID)
	notification := map[string]interface{}{
		"type":    "kyc",
		"status":  status,
		"message": message,
		"time":    time.Now().Unix(),
	}
	
	notificationJSON, _ := json.Marshal(notification)
	s.redis.LPush(s.redis.Context(), key, notificationJSON)
	s.redis.Expire(s.redis.Context(), key, 7*24*time.Hour)
	
	return nil
}

// KYCHandler handles KYC-related requests
type KYCHandler struct {
	db               *database.PostgresDB
	notificationSvc *NotificationService
}

// NewKYCHandler creates a new KYC handler
func NewKYCHandler(db *database.PostgresDB, redisClient *redis.Client) *KYCHandler {
	return &KYCHandler{
		db:               db,
		notificationSvc: NewNotificationService(redisClient),
	}
}

// ListKYC lists all KYC applications
func (h *KYCHandler) ListKYC(c *gin.Context) {
	page := c.DefaultQuery("page", "1")
	pageSize := c.DefaultQuery("page_size", "20")
	status := c.Query("status")
	level := c.Query("level")

	pageInt, _ := strconv.Atoi(page)
	pageSizeInt, _ := strconv.Atoi(pageSize)

	var applications []models.KYCApplication
	var total int64

	query := h.db.Model(&models.KYCApplication{}).Preload("User")

	if status != "" {
		query = query.Where("status = ?", status)
	}
	if level != "" {
		levelInt, _ := strconv.Atoi(level)
		query = query.Where("level = ?", levelInt)
	}

	query.Count(&total)

	offset := (pageInt - 1) * pageSizeInt
	query = query.Offset(offset).Limit(pageSizeInt).Order("created_at DESC")

	if err := query.Find(&applications).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch KYC applications"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"data":        applications,
		"total":       total,
		"page":        pageInt,
		"page_size":   pageSizeInt,
		"total_pages": (total + int64(pageSizeInt) - 1) / int64(pageSizeInt),
	})
}

// GetKYC gets a KYC application by ID
func (h *KYCHandler) GetKYC(c *gin.Context) {
	kycID := c.Param("id")

	var kyc models.KYCApplication
	if err := h.db.Preload("User").First(&kyc, kycID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "KYC application not found"})
		return
	}

	c.JSON(http.StatusOK, kyc)
}

// ApproveKYC approves a KYC application
func (h *KYCHandler) ApproveKYC(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	kycID := c.Param("id")

	var req struct {
		Notes string `json:"notes"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		req.Notes = ""
	}

	var kyc models.KYCApplication
	if err := h.db.Preload("User").First(&kyc, kycID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "KYC application not found"})
		return
	}

	if kyc.Status != "pending" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "KYC application is not pending"})
		return
	}

	now := time.Now()

	// Update KYC application
	if err := h.db.Model(&kyc).Updates(map[string]interface{}{
		"status":     "approved",
		"reviewed_at": now,
		"reviewed_by": adminID,
		"notes":      req.Notes,
	}).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to approve KYC"})
		return
	}

	// Update user KYC status
	if err := h.db.Model(&models.User{}).Where("id = ?", kyc.UserID).Updates(map[string]interface{}{
		"kyc_status":     "level" + strconv.Itoa(kyc.Level),
		"kyc_level":      kyc.Level,
		"kyc_verified_at": now,
	}).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update user KYC status"})
		return
	}

	// Log activity
	logAdminActivity(h.db, adminID, "approve_kyc", "kyc", kycID, "KYC approved at level "+strconv.Itoa(kyc.Level), c.ClientIP(), c.Request.UserAgent())

	// Send notification to user
	notificationMsg := fmt.Sprintf("Your KYC application has been approved at Level %d. You now have access to higher withdrawal limits and additional features.", kyc.Level)
	_ = h.notificationSvc.SendKYCNotification(kyc.UserID, "approved", notificationMsg)

	c.JSON(http.StatusOK, gin.H{"message": "KYC approved successfully"})
}

// RejectKYC rejects a KYC application
func (h *KYCHandler) RejectKYC(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	kycID := c.Param("id")

	var req struct {
		Reason string `json:"reason" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Rejection reason is required"})
		return
	}

	var kyc models.KYCApplication
	if err := h.db.Preload("User").First(&kyc, kycID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "KYC application not found"})
		return
	}

	if kyc.Status != "pending" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "KYC application is not pending"})
		return
	}

	now := time.Now()

	// Update KYC application
	if err := h.db.Model(&kyc).Updates(map[string]interface{}{
		"status":            "rejected",
		"reviewed_at":       now,
		"reviewed_by":       adminID,
		"rejection_reason":  req.Reason,
	}).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to reject KYC"})
		return
	}

	// Update user KYC status
	if err := h.db.Model(&models.User{}).Where("id = ?", kyc.UserID).Updates(map[string]interface{}{
		"kyc_status":           "rejected",
		"kyc_rejection_reason": req.Reason,
	}).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update user KYC status"})
		return
	}

	// Log activity
	logAdminActivity(h.db, adminID, "reject_kyc", "kyc", kycID, "KYC rejected: "+req.Reason, c.ClientIP(), c.Request.UserAgent())

	// Send notification to user
	notificationMsg := fmt.Sprintf("Your KYC application has been rejected. Reason: %s. Please submit a new application with the required documents.", req.Reason)
	_ = h.notificationSvc.SendKYCNotification(kyc.UserID, "rejected", notificationMsg)

	c.JSON(http.StatusOK, gin.H{"message": "KYC rejected"})
}

// GetKYCStats gets KYC statistics
func (h *KYCHandler) GetKYCStats(c *gin.Context) {
	var stats struct {
		Pending  int64 `json:"pending"`
		Approved int64 `json:"approved"`
		Rejected int64 `json:"rejected"`
		Total    int64 `json:"total"`
	}

	h.db.Model(&models.KYCApplication{}).Where("status = ?", "pending").Count(&stats.Pending)
	h.db.Model(&models.KYCApplication{}).Where("status = ?", "approved").Count(&stats.Approved)
	h.db.Model(&models.KYCApplication{}).Where("status = ?", "rejected").Count(&stats.Rejected)
	h.db.Model(&models.KYCApplication{}).Count(&stats.Total)

	c.JSON(http.StatusOK, stats)
}
